lib/fileset/default.nix

```diff
@@ -278,4 +281,115 @@ If a directory does not recursively contain any file, it is omitted from the sto
         _unionMany
       ];
 
+  /*
+    Incrementally evaluate and trace a file set in a pretty way.
+    This function is only intended for debugging purposes.
+    The exact tracing format is unspecified and may change.
+
+    This function takes a final argument to return.
+    In comparison, [`traceVal`](#function-library-lib.fileset.traceVal) returns
+    the given file set argument.
+
+    This variant is useful for tracing file sets in the Nix repl.
+
+    Type:
+      trace :: { } -> FileSet -> Any -> Any
+
+    Example:
+      trace {} (unions [ ./Makefile ./src ./tests/run.sh ]) null
+      =>
+      trace: /home/user/src/myProject
+      trace: - Makefile (regular)
+      trace: - src (all files in directory)
+      trace: - tests
+      trace:   - run.sh (regular)
+      null
+  */
+  trace =
```

Comment:
The main use case for `traceVal` is when you have something like
```nix
toSource {
  root = ./.;
  fileset = unions [
    ./Makefile
    ./srcs
  ];
}
```

And you want to trace the `fileset` argument. To use trace here without duplication requires a new variable, which is a bit of a hassle:
```nix
toSource {
  root = ./.;
  fileset =
    let
      fileset = unions [
        ./Makefile
        ./srcs
      ];
    in
    trace {} fileset fileset;
}
```

Comparatively with `traceVal` it's almost a one-line change: 
```nix
toSource {
  root = ./.;
  fileset = traceVal {} (unions [
    ./Makefile
    ./srcs
  ]);
}
```

---

Furthermore, it's useful if you want to trace some intermediate file set like here:
```nix
intersect ./.
  (difference ../. (unions [
    (fileFilter (file: file.name == "default.nix") ./.)
    ./Makefile
  ])
```

Say you want to know what files are contained in the result of the file filter. Without `traceVal`:

```nix
intersect ./.
  (difference ../. (unions [
    (
      let
        result = fileFilter (file: file.name == "default.nix") ./.;
      in
      trace {} result result
    )
    ./Makefile
  ])
```

With `traceVal`:
```nix
intersect ./.
  (difference ../. (unions [
    (traceVal {} (fileFilter (file: file.name == "default.nix") ./.))
    ./Makefile
  ])
```

Reply:
Right, therefore I was asking whether it would make sense to remove what's currently `trace` and rename `traceVal` to `trace`. What's currently `traceVal` seems to be much more obviously useful.

Reply:
Ah I see! The main use case for the current `trace` that I have is really the nix repl:

```
nix-repl> fileset.trace {} (union ./default.nix ./lib)
trace: /home/infinisil/src/nixpkgs
trace: - default.nix (regular)
trace: - lib (all files in directory)
«lambda @ /home/infinisil/src/nixpkgs/lib/fileset/default.nix:334:10»
```

Whereas for `traceVal`, it would try to evaluate the result, which is rather unsighty:
```
nix-repl> fileset.traceVal {} (union ./default.nix ./lib) 
trace: /home/infinisil/src/nixpkgs
trace: - default.nix (regular)
trace: - lib (all files in directory)
{ _internalBase = /home/infinisil/src/nixpkgs; _internalBaseComponents = [ ... ]; _internalBaseRoot = /; _internalTree = { ... }; _internalVersion = 2; _noEval = «error: error:
       … while calling the 'throw' builtin

         at /home/infinisil/src/nixpkgs/lib/fileset/internal.nix:117:17:

          116|       # It also allows evaluating it strictly up to this error, which could be useful
          117|       _noEval = throw ''
             |                 ^
          118|         lib.fileset: Directly evaluating a file set is not supported. Use `lib.fileset.toSource` to turn it into a usable source instead.'';

       error: lib.fileset: Directly evaluating a file set is not supported. Use `lib.fileset.toSource` to turn it into a usable source instead.»; _type = "fileset"; }
```

But perhaps this should be fixed in the nix repl by having support for some optional  attribute to use for pretty-printing values, e.g. `_nixReplShow = self: "..."`. Though we don't have that right now.

I do think the names should stay as they are though, because we also have `lib.debug.traceVal` and `builtins.trace` with matching semantics. And considering that, the current `fileset.trace` semantics is the only function that would fit into its name, might as well just have it. Furthermore, there's barely any cost to providing two trace variants, since almost all of their test cases are shared, and they use the same internal code.

I vote to keep it :)

Reply:
Thanks for the explanation.